### PR TITLE
Throw an exception on invalid cache configuration

### LINKED ISSUE #14985: Failure to setup second Memcached instance on another port
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.x.
* Platform and Target: Ubuntu, MySQL, Apache.

### What you did
1. Started another memcached instance on port 11213 by running `memcached -p 11213 -d` in the console.
2. Configured `'_cake_core_'` cache to use that instance by using `'servers' => ['localhost:11213'],`

### What happened
`'_cake_core_'` used the default memcached on port 11211 instead 11213.

The line [157](https://github.com/cakephp/cakephp/blob/master/src/Cache/Engine/MemcachedEngine.php#L157) `if (!$this->_Memcached->addServers($servers)) {` seems to be never executed because `$this->_Memcached->getServerList()` on the line [148](https://github.com/cakephp/cakephp/blob/master/src/Cache/Engine/MemcachedEngine.php#L148) in MemcachedEngine always  returns `true`.

### What you expected to happen
I would expect `'_cake_core_'`  to use 'localhost:11213' memcached.

### PR DESCRIPTION
**Pull request by markstory**
* Add test case for the memcached config on the second port, #14985
* Port changes from #14986 to 4.x

Fixes #14985

### CODE REVIEW

File: src/Cache/Engine/MemcachedEngine.php [hunk 1]
```diff
@@ -145,7 +145,25 @@ public function init(array $config = []): bool
         }
         $this->_setOptions();
 
-        if (count($this->_Memcached->getServerList())) {
+        $serverList = $this->_Memcached->getServerList();
+        if (count($serverList)) {
```
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:149]:
```suggestion
        if ($serverList) {
```

File: src/Cache/Engine/MemcachedEngine.php [hunk 2]
```diff
@@ -145,7 +145,25 @@ public function init(array $config = []): bool
...
+            if ($this->_config['persistent'] !== false) {
```
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:150]:
```suggestion
        if ($serverList && $this->_config['persistent'] !== false) {
```

File: src/Cache/Engine/MemcachedEngine.php (hunk 1 above)
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:149]:
```suggestion
        if ($serverList) {
```

File: src/Cache/Engine/MemcachedEngine.php [hunk 3]
```diff
@@ -145,7 +145,25 @@ public function init(array $config = []): bool
...
+                $actualServers = [];
+                foreach ($serverList as $server) {
+                    $actualServers[] = $server['host'] . ':' . $server['port'];
+                }
+                unset($server);
```
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:155]: why is this unset?

File: src/Cache/Engine/MemcachedEngine.php (hunk 2 above)
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:150]:
```suggestion
            if ($this->_config['persistent']) {
```
null shouldn't be persistent, right?

Comment (othercorey) [reply src/Cache/Engine/MemcachedEngine.php:150]: Can we use `$this->_Memcached->isPersistent()` here?

File: src/Cache/Engine/MemcachedEngine.php [hunk 4]
```diff
@@ -145,7 +145,25 @@ public function init(array $config = []): bool
...
+                sort($actualServers);
+                $configuredServers = $this->_config['servers'];
+                sort($configuredServers);
+                if ($actualServers !== $configuredServers) {
+                    $message = 'Invalid cache configuration. Multiple persistent cache configurations are detected' .
+                        ' with different `servers` values. `servers` values for persistent cache configurations' .
+                        ' must be the same when using the same persistence id.';
+                    throw new InvalidArgumentException($message);
+                }
```
Comment (othercorey) [inline src/Cache/Engine/MemcachedEngine.php:164]:
```suggestion
                foreach ($serverList as $server) {
                    if (!in_array($server['host'] . ':' . $server['port'], $this->_config['server'], true)) {
                        throw new InvalidArgumentException(
                            'Invalid cache configuration. Multiple persistent cache configurations are detected' .
                            ' with different `servers` values. `servers` values for persistent cache configurations' .
                            ' must be the same when using the same persistence id.'
                        );
                    }
                }
                $actualServers = [];
                foreach ($serverList as $server) {
                    $actualServers[] = $server['host'] . ':' . $server['port'];
                }
                unset($server);
                sort($actualServers);
                $configuredServers = $this->_config['servers'];
                sort($configuredServers);
                if ($actualServers !== $configuredServers) {
                    $message = 'Invalid cache configuration. Multiple persistent cache configurations are detected' .
                        ' with different `servers` values. `servers` values for persistent cache configurations' .
                        ' must be the same when using the same persistence id.';
                    throw new InvalidArgumentException($message);
                }
```

Comment (othercorey) [reply src/Cache/Engine/MemcachedEngine.php:164]: I don't know if that's the best exception message, but I'm not familiar with it.

File: tests/TestCase/Cache/Engine/MemcachedEngineTest.php [hunk 5]
```diff
@@ -374,6 +374,32 @@ public function testSaslAuthException()
         $MemcachedEngine->init($config);
     }
 
+    /**
+     * testConfigDifferentPorts method
+     *
+     * @expectedException InvalidArgumentException
```
Comment (othercorey) [inline tests/TestCase/Cache/Engine/MemcachedEngineTest.php:380]: I believe this tags are deprecated. Use $this->expectException().

File: src/Cache/Engine/MemcachedEngine.php (hunk 2 above)
Comment (markstory) [reply src/Cache/Engine/MemcachedEngine.php:150]: We might be able to.

**Pull request merged**